### PR TITLE
Use registered strings for NVTX. Add more NVTX annotations.

### PR DESCRIPTION
**Pull request by kingcrimsontianyu**
This PR makes some improvements on NVTX annotations in KvikIO, including:
- Improve NVTX range annotation by registering the range name string (which reduces profiling-time overhead) and making the registered string static (which further reduces the overhead).
- Separate the responsibility of "scoped range" from "function range" to improve clarity. Previously variadic macro `KVIKIO_NVTX_FUNC_RANGE(...)` is used for both the function range and arbitrary scoped range. However, the macro is somewhat a misnomer for the latter case, where `__func__` is not used in the message. This PR introduces `KVIKIO_NVTX_SCOPED_RANGE(msg, val)` for the latter case.
- Add NVTX marker macro `KVIKIO_NVTX_MARKER(message, payload)` to annotate an instantaneous event. One use case is to characterize asynchronous I/O operations (where the scoped range in a function does not apply): Once the profile is collected, the `nsys-rep` file can be exported, for instance, as a `sqlite`, and the marker data can be used to make a I/O-size histogram. As an example, for `PARQUET_READER_NVBENCH---parquet_read_io_compression`, the following I/O statistics can be obtained.
This histogram is generated using the Python script `python/kvikio/examples/kvikio_stat.py`. PR link: TBD...

```
FileHandle::pread()
    Bins                 ...... Count
    [     0 B,      2 B) ...... 0
    [     2 B,      4 B) ...... 0
    [     4 B,      8 B) ...... 0
    [     8 B,     16 B) ...... 0
    [    16 B,     32 B) ...... 0
    [    32 B,     64 B) ...... 0
    [    64 B,    128 B) ...... 0
    [   128 B,    256 B) ...... 0
    [   256 B,    512 B) ...... 0
    [   512 B,   1024 B) ...... 0
    [  1024 B,     2 KB) ...... 0
    [    2 KB,     4 KB) ...... 0
    [    4 KB,     8 KB) ...... 0
    [    8 KB,    16 KB) ...... 0
    [   16 KB,    32 KB) ...... 0
    [   32 KB,    64 KB) ...... 0
    [   64 KB,   128 KB) ...... 0
    [  128 KB,   256 KB) ...... 45
    [  256 KB,   512 KB) ...... 30
    [  512 KB,  1024 KB) ...... 15
    [ 1024 KB,     2 MB) ...... 120
    [    2 MB,     4 MB) ...... 150
    [    4 MB,     8 MB) ...... 60
    [    8 MB,    16 MB) ...... 0
    [   16 MB,    32 MB) ...... 0
    [   32 MB,    64 MB) ...... 0
    [   64 MB,   128 MB) ...... 0
    [  128 MB,   256 MB] ...... 30



FileHandle::pwrite()
    Bins                 ...... Count
    [     0 B,      2 B) ...... 0
    [     2 B,      4 B) ...... 0
    [     4 B,      8 B) ...... 0
    [     8 B,     16 B) ...... 0
    [    16 B,     32 B) ...... 0
    [    32 B,     64 B) ...... 0
    [    64 B,    128 B) ...... 0
    [   128 B,    256 B) ...... 0
    [   256 B,    512 B) ...... 0
    [   512 B,   1024 B) ...... 0
    [  1024 B,     2 KB) ...... 0
    [    2 KB,     4 KB) ...... 0
    [    4 KB,     8 KB) ...... 0
    [    8 KB,    16 KB) ...... 0
    [   16 KB,    32 KB) ...... 0
    [   32 KB,    64 KB) ...... 0
    [   64 KB,   128 KB) ...... 0
    [  128 KB,   256 KB) ...... 8
    [  256 KB,   512 KB) ...... 8
    [  512 KB,  1024 KB) ...... 19
    [ 1024 KB,     2 MB) ...... 24
    [    2 MB,     4 MB) ...... 4
    [    4 MB,     8 MB) ...... 0
    [    8 MB,    16 MB) ...... 0
    [   16 MB,    32 MB) ...... 0
    [   32 MB,    64 MB) ...... 0
    [   64 MB,   128 MB) ...... 0
    [  128 MB,   256 MB] ...... 2



posix_device_read()
    Bins                 ...... Count
    [     0 B,      2 B) ...... 0
    [     2 B,      4 B) ...... 0
    [     4 B,      8 B) ...... 0
    [     8 B,     16 B) ...... 0
    [    16 B,     32 B) ...... 0
    [    32 B,     64 B) ...... 0
    [    64 B,    128 B) ...... 0
    [   128 B,    256 B) ...... 0
    [   256 B,    512 B) ...... 0
    [   512 B,   1024 B) ...... 0
    [  1024 B,     2 KB) ...... 0
    [    2 KB,     4 KB) ...... 0
    [    4 KB,     8 KB) ...... 0
    [    8 KB,    16 KB) ...... 0
    [   16 KB,    32 KB) ...... 0
    [   32 KB,    64 KB) ...... 0
    [   64 KB,   128 KB) ...... 0
    [  128 KB,   256 KB) ...... 75
    [  256 KB,   512 KB) ...... 60
    [  512 KB,  1024 KB) ...... 15
    [ 1024 KB,     2 MB) ...... 150
    [    2 MB,     4 MB] ...... 1710



posix_device_write()
    Bins                 ...... Count
    [     0 B,      2 B) ...... 0
    [     2 B,      4 B) ...... 0
    [     4 B,      8 B) ...... 0
    [     8 B,     16 B) ...... 0
    [    16 B,     32 B) ...... 0
    [    32 B,     64 B) ...... 0
    [    64 B,    128 B) ...... 0
    [   128 B,    256 B) ...... 0
    [   256 B,    512 B) ...... 0
    [   512 B,   1024 B) ...... 0
    [  1024 B,     2 KB) ...... 0
    [    2 KB,     4 KB) ...... 0
    [    4 KB,     8 KB) ...... 0
    [    8 KB,    16 KB) ...... 0
    [   16 KB,    32 KB) ...... 0
    [   32 KB,    64 KB) ...... 0
    [   64 KB,   128 KB) ...... 0
    [  128 KB,   256 KB) ...... 8
    [  256 KB,   512 KB) ...... 8
    [  512 KB,  1024 KB) ...... 19
    [ 1024 KB,     2 MB) ...... 24
    [    2 MB,     4 MB] ...... 104
```

### CODE REVIEW

File: cpp/include/kvikio/utils.hpp [hunk 1]
```diff
@@ -287,15 +287,27 @@ struct libkvikio_domain {
...
-    {                                                         \
-      msg, nvtx3::payload { convert_to_64bit(val) }           \
-    }                                                         \
+#define KVIKIO_NVTX_FUNC_RANGE_2(msg, val)                                      \
+  nvtx3::scoped_range_in<libkvikio_domain> CONCAT(_kvikio_nvtx_range, __LINE__) \
+  {                                                                             \
+    nvtx3::event_attributes                                                     \
+    {                                                                           \
+      REGISTER_STRING(msg), nvtx3::payload { convert_to_64bit(val) }            \
+    }                                                                           \
```
Comment (kingcrimsontianyu) [inline cpp/include/kvikio/utils.hpp:310]: My initial change was simpler:
```
#define KVIKIO_NVTX_FUNC_RANGE_2(msg, val)                                               \
  static nvtx3::registered_string_in<libkvikio_domain> CONCAT(a_reg_str, __LINE__){msg}; \
  nvtx3::scoped_range_in<libkvikio_domain> CONCAT(_kvikio_nvtx_range, __LINE__)          \
  {                                                                                      \
    nvtx3::event_attributes                                                              \
    {                                                                                    \
      CONCAT(a_reg_str, __LINE__), nvtx3::payload { convert_to_64bit(val) }              \
    }                                                                                    \
  }
```
However, I was not sure if the standard specifies that `__LINE__` would remain the same for the two `CONCAT(a_reg_str, __LINE__)`s (although the code did compile). Hence the lambda workaround.

File: cpp/include/kvikio/utils.hpp [hunk 2]
```diff
@@ -330,4 +347,29 @@ struct libkvikio_domain {
...
+ * ```
+ * std::future<void> some_function(){
+ *     size_t io_size{2077};
+ *     KVIKIO_NVTX_MARKER("I/O operation", io_size);
+ *     perform_async_io_operation(io_size);
+ *     ...
+ * }
+ * ```
+ */
+#ifdef KVIKIO_CUDA_FOUND
```
Comment (bdice) [inline cpp/include/kvikio/utils.hpp:367]: NVTX can be used without CUDA in host-only code. We can disable this if we don't care about profiling no-CUDA code paths, though.

Comment (madsbk) [reply cpp/include/kvikio/utils.hpp:367]: I think the guard is needed, we only build with NVTX support when `KVIKIO_CUDA_FOUND` is defined: https://github.com/rapidsai/kvikio/blob/branch-24.12/cpp/CMakeLists.txt#L92

**Review by madsbk — APPROVED** [review]
Thanks @kingcrimsontianyu, looks good to me. 

PS: the current CI error in `test_read_access(s3_base)` is not related to your changes. I will work on  a fix tomorrow,

**Comment by madsbk** [comment]
@kingcrimsontianyu could you add a description in the docs on how to generate the histograms? They are very cool!

File: cpp/include/kvikio/utils.hpp (hunk 2 above)
Comment (bdice) [reply cpp/include/kvikio/utils.hpp:367]: Okay, sounds good. It is _possible_ to use NVTX without CUDA, but I'm fine with this outcome given the lack of CMake support for that.

**Comment by kingcrimsontianyu** [comment]
@madsbk Sure! The histogram is generated by a Python script. I can make a separate PR and update the doc. Is `python/kvikio/examples` the right place for that script?

**Comment by madsbk** [comment]
> @madsbk Sure! The histogram is generated by a Python script. I can make a separate PR and update the doc. Is `python/kvikio/examples` the right place for that script?

Yes, that would be great!

**Review by madsbk — CHANGES_REQUESTED** [review]

File: cpp/include/kvikio/utils.hpp [hunk 3]
```diff
@@ -287,17 +287,34 @@ struct libkvikio_domain {
   static constexpr char const* name{"libkvikio"};
 };
 
+// Macro to concatenate two tokens x and y.
+#define CONCAT_HELPER(x, y) x##y
+#define CONCAT(x, y)        CONCAT_HELPER(x, y)
+
+// Macro to create a static, registered string that will not have a name conflict with any
+// registered string defined in the same scope.
+#define REGISTER_STRING(msg)                                               \
```
[the remark this answers is not in the document]
Comment (kingcrimsontianyu) [reply cpp/include/kvikio/utils.hpp]: Good idea. Completed.

**Review by madsbk — APPROVED** [review]
Nice work @kingcrimsontianyu

**Pull request merged**